docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,341 @@
+        assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+    })
+    .expect("can spawn a thread");
+
+// Begin running the instance, which will be terminated remotely by the KillSwitch.
+match inst.run("infinite_loop", &[]) {
+    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
+        // the result of a guest that was remotely terminated (a KillSwitch at work)
+    }
+    res => panic!("unexpected result: {:?}", res),
+}
+
+// wait for the KillSwitch-firing thread to complete
+t.join().unwrap();
+```
+
+## Implementation
+
+As this section discusses implementation details of `lucet-runtime`, it
+will refer to structures and enums that are unexported.  Since there are
+no `lucet-runtime` paths for these items, this section will occasionally
+use fully-qualified paths to the `lucet_runtime_internals` path where.
+they are originally defined.
+
+As much as is possible, `KillSwitch` tries to be self-contained; no members are
+public, and it tries to avoid leaking details of its state-keeping into public
+interfaces. Currently, `lucet-runtime` is heavily dependent on POSIX
+thread-directed signals to implement guest timeouts - for non-POSIX platforms
+alternate implementations may be plausible, but need careful consideration of
+the race conditions that can arise from other platform-specific functionality.
+
+`KillSwitch` fundamentally relies on two piece of state for safe operation,
+which are encapsulated in a `KillState` held by the `lucet_runtime::Instance`
+it terminates:
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`. This is kept in a `Mutex` since in
+  many cases it will need to be accessed either by a `KillSwitch` or
+  `KillState`, and for the duration either are considering the domain, it must
+  block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
+
+The astute observer may notice that `Lucet` contains both an instance `Domain`
+and an instance `State`. These discuss different aspects of the instance's
+lifecycle: `Domain` describes what the instance is doing right now, where
+`State` describes the state of the instance as visible outside
+`lucet_runtime`. This is fundamentally the reason why `State` does not and
+cannot describe a "making a hostcall" variant - this is not something
+`lucet_runtime` wants to expose, and at the moment we don't intend to make any
+commitments about being able to query instance-internal state like this.
+Additionally, this is why `Domain` only expresses that an instance has
+_stopped_ running, not if it successfully or unsuccessfully exited. TODO:
+there's even an argument that `Terminated` vs `Cancelled` is a stronger
+distinction than `Domain` needs to make!
+
+### Termination Mechanisms
+
+At a high level,`KillSwitch` picks one of several mechanisms to terminate an
+instance.  The choice of which mechanism depends on what kind of code is
```

Comment:
It seems that the choice of mechanism depends not just on what kind of code is running in the guest, but also whether or not it's pending.  It might be more precise to leave out the first part of this sentence and just offer that "The mechanism selected depends on `instance::execution::Domain`.